hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -136,20 +174,101 @@ private IndexResolver() {
+            List<Boolean> ascs = buildFieldDirections(index, true);
+            RelNode relAscending = createIndexScan(scan, distribution, index, conjunctions, candidates, ascs);
+
+            if (relAscending != null) {
+                RelCollation relAscCollation = getCollation(relAscending);
+                if (fullScanRelsMap.containsKey(relAscCollation)) {
+                    // The fulls scan has the same collation, exclude it
+                    fullScanRelsMap.remove(relAscCollation);
+                }
+
+                rels.add(relAscending);
 
-            if (rel != null) {
-                rels.add(rel);
+                if (relAscCollation.getFieldCollations().size() > 0) {
+                    RelNode relDescending = replaceCollationDirection(relAscending, DESCENDING);
+                    rels.add(relDescending);
+
+                    RelCollation relDescCollation = getCollation(relDescending);
+                    if (fullScanRelsMap.containsKey(relDescCollation)) {
+                        // The fulls scan has the same collation, exclude it
+                        fullScanRelsMap.remove(relDescCollation);
+                    }
+                }
             }
         }
 
+        rels.addAll(fullScanRelsMap.values());
         return rels;
     }
 
+    private static RelCollation getCollation(RelNode rel) {
+        return rel.getTraitSet().getTrait(RelCollationTraitDef.INSTANCE);
+    }
+
+    /**
+     * Replaces a direction in the collation trait of the rel
+     * @param rel       the rel
+     * @param direction the collation
+     * @return the rel with changed collation
+     */
+    private static RelNode replaceCollationDirection(RelNode rel, Direction direction) {
+        RelCollation collation = rel.getTraitSet().getTrait(RelCollationTraitDef.INSTANCE);
+
+        List<RelFieldCollation> newFields = new ArrayList<>(collation.getFieldCollations().size());
+        for (RelFieldCollation fieldCollation : collation.getFieldCollations()) {
+            RelFieldCollation newFieldCollation = new RelFieldCollation(fieldCollation.getFieldIndex(), direction);
+            newFields.add(newFieldCollation);
+        }
+
+        RelCollation newCollation = RelCollations.of(newFields);
+        RelTraitSet traitSet = rel.getTraitSet();
+        traitSet = OptUtils.traitPlus(traitSet, newCollation);
+
+        return rel.copy(traitSet, rel.getInputs());
+    }
+
+    /**
+     * Filters out index scans which collation is covered (prefix based) by another index scan in the rels.
+     *
+     * @param rels the list of index scans
+     * @return a filtered out map of collation to rel
+     */
+    @SuppressWarnings("checkstyle:NPathComplexity")
+    private static Map<RelCollation, RelNode> excludeCoveredCollations(List<RelNode> rels) {
+        // Order the index scans based on their collation
+        TreeMap<RelCollation, RelNode> relsTreeMap = new TreeMap<>(new RelCollationComparator());
```

Comment:
`RelCollationComparator` could be a singleton.

Reply:
Addressed in https://github.com/hazelcast/hazelcast/pull/18013/commits/0d7ec9ad4314ac80c0cf1173689c0f7b98513538